backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/modal/BpkModalState.kt

```diff
@@ -32,12 +33,22 @@ class BpkModalState internal constructor(
     internal val isVisible: MutableTransitionState<Boolean>,
 ) {
     suspend fun show() {
-        isVisible.targetState = true
-        delay(ModalAnimationDurationMs.toLong())
+        animateState(true)
     }
 
     suspend fun hide() {
-        isVisible.targetState = false
-        delay(ModalAnimationDurationMs.toLong())
+        animateState(false)
+    }
+
+    private suspend fun animateState(newVisibility: Boolean) {
+        isVisible.targetState = newVisibility
+        if (isVisible.currentState == newVisibility) {
+            return
+        }
+        withContext(Dispatchers.IO) {
+            while (!isVisible.isIdle) {
```

Comment:
we now just wait until the animation is no longer idle - this means that no matter the animation speed we'll run this function as long as the animation runs

Reply:
IIRC, it's kind of spin lock here, and will busy looping till the animation end (in most device will took a cpu core  ~300ms), this look a bit expensive to just for the animation?

How about put a `delay` inside the `while`:
```
while (!isVisible.isIdle) {
    delay(33L)  // 33ms is a common frame interval for most devices
}
```

Reply:
just looping: 
<img width="789" alt="image" src="https://github.com/Skyscanner/backpack-android/assets/1937740/61d3f85c-3151-4382-a136-21dcb611cac7">

looping with delay 8ms:
<img width="639" alt="image" src="https://github.com/Skyscanner/backpack-android/assets/1937740/718ae1d8-ee23-4b01-a5e9-2dec505c09a6">

from the profiler, we can see that the peak cpu usage is dropped after add a `delay(8L)`

Reply:
yeah, gonna create a PR about this

Reply:
hmm yes you might be right - i think i'd like to spend some more time on this later on to find a good solution for this tbh - with the delay we risk there being issues as well as currently the idle check will always happen at the same time (or rather there won't be an ui update in between) - this means we can rely on this callback to be called when the modal/viewmodel scope is still alive. as soon as we add a delay we risk a race condition where sometimes depending on the delay the modal is already dismissed/the viewmodel gone and our bug is back

Reply:
> hmm yes you might be right - i think i'd like to spend some more time on this later on to find a good solution for this tbh - with the delay we risk there being issues as well as currently the idle check will always happen at the same time (or rather there won't be an ui update in between) - this means we can rely on this callback to be called when the modal/viewmodel scope is still alive. as soon as we add a delay we risk a race condition where sometimes depending on the delay the modal is already dismissed/the viewmodel gone and our bug is back

emmmm, i think you're right, if we add a `delay` here, then this couroutine will be switch out of thread, at that time, if a recompose was trigged, the couroutine job will be close.
maybe we can keep the current source code (without `delay`), so that we can fix the bug, then create another PR to optimize this spin lock.